core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -872,6 +872,25 @@ class DAGScheduler(
     // will be posted, which should always come after a corresponding SparkListenerStageSubmitted
     // event.
     stage.latestInfo = StageInfo.fromStage(stage, Some(partitionsToCompute.size))
+    val taskIdToLocations = try {
+      stage match {
+        case s: ShuffleMapStage =>
+          partitionsToCompute.map { id => (id, getPreferredLocs(stage.rdd, id))}.toMap
+        case s: ResultStage =>
+          val job = s.resultOfJob.get
+          partitionsToCompute.map { id =>
+            val p = job.partitions(id)
+            (id, getPreferredLocs(stage.rdd, p))
+          }.toMap
+      }
+    } catch {
+      case NonFatal(e) =>
+        abortStage(stage, s"Task creation failed: $e\n${e.getStackTraceString}")
+        runningStages -= stage
+        return
+    }
+    stage.latestInfo.taskLocalityPreferences = Some(taskIdToLocations.values.toSeq)
```

Comment:
Does it make sense to move this initialization code to Stage.fromInfo? Then the taskLocalityPreferences doesn't need to be an option, and it's also more consistent with the way the rest of that object is constructed.